### PR TITLE
Add Algorithm Euler method

### PR DESCRIPTION
**Pull request by algobytewise**
Please include a summary of the change. Please also include relevant motivation and context (if applicable).

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have made corresponding changes to the comments
- [ ] I have made corresponding changes to the README.md

### CODE REVIEW

File: Algorithms.Tests/Numeric/EulerMethodTest.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+using Algorithms.Numeric;
+using NUnit.Framework;
+using System.Collections.Generic; 
+
+namespace Algorithms.Tests.Numeric
+{
+    public static class EulerMethodTest
+    {
+        [Test]
+        public static void Test1()
```
Comment (siriak) [inline Algorithms.Tests/Numeric/EulerMethodTest.cs:11]: Please give descriptive names to the tests

File: Algorithms/Numeric/EulerMethod.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        public static List<double[]> EulerFull(
+                double xStart,
+                double xEnd,
+                double stepSize,
+                double yStart,
+                Func<double, double, double> differentialEquation)
+        {
+            if (xStart >= xEnd)
+            {
+              throw new ArgumentOutOfRangeException("xEnd", "xEnd should be greater than xStart");
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:61]:
```suggestion
              throw new ArgumentOutOfRangeException(nameof(xEnd), $"{nameof(xEnd)} should be greater than {nameof(xStart)}");
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            }
+
+            if (stepSize <= 0)
+            {
+              throw new ArgumentOutOfRangeException("stepSize", "stepSize should be greater than zero");
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:66]:
```suggestion
              throw new ArgumentOutOfRangeException(nameof(stepSize), $"{nameof(stepSize)} should be greater than zero");
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        /// <summary>
+        /// Loops through all the steps until xEnd is reached, adds a point for each step and then
+        /// returns all the points.
+        /// </summary>
+        /// <param name="xStart">First x-value.</param>
+        /// <param name="xEnd">Last x-value.</param>
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yStart">First y-value.</param>
+        /// <param name="differentialEquation">The differential equation to be solved.</param>
+        /// <returns>The points constituting the solution of the differential equation.</returns>
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:51]:
```suggestion
        /// <returns>The solution of the Cauchy problem.</returns>
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        /// <summary>
+        /// Loops through all the steps until xEnd is reached, adds a point for each step and then
+        /// returns all the points.
+        /// </summary>
+        /// <param name="xStart">First x-value.</param>
+        /// <param name="xEnd">Last x-value.</param>
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yStart">First y-value.</param>
+        /// <param name="differentialEquation">The differential equation to be solved.</param>
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:50]:
```suggestion
        /// <param name="differentialEquation">The right hand side of the differential equation.</param>
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+            double yNext = yCurrent + stepSize * differentialEquation(xCurrent, yCurrent);
+            return yNext;
+        }
+
+        /// <summary>
+        /// Loops through all the steps until xEnd is reached, adds a point for each step and then
+        /// returns all the points.
+        /// </summary>
+        /// <param name="xStart">First x-value.</param>
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:46]:
```suggestion
        /// <param name="xStart">Initial conditions x-value.</param>
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        /// <param name="xEnd">Last x-value.</param>
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yStart">First y-value.</param>
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:49]:
```suggestion
        /// <param name="yStart">Initial conditions y-value.</param>
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        /// <summary>
+        /// Calculates the next y-value based on the current value of x, y and the stepSize.
+        /// </summary>
+        /// <param name="xCurrent">Current x-value.</param>
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yCurrent">Current y-value.</param>
+        /// <param name="differentialEquation">The differential equation to be solved.</param>
+        /// <returns>The next y-value.</returns>
+        public static double EulerStep(
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:27]:
```suggestion
        private static double EulerStep(
```

Comment (siriak) [reply Algorithms/Numeric/EulerMethod.cs:27]: No need to make it public

File: Algorithms/Numeric/EulerMethod.cs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                double xCurrent,
+                double stepSize,
+                double yCurrent,
+                Func<double, double, double> differentialEquation)
+        {
+            if (stepSize <= 0)
+            {
+              throw new ArgumentOutOfRangeException("stepSize", "stepSize should be greater than zero");
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:35]:
```suggestion
              throw new ArgumentOutOfRangeException(nameof(stepSize), $"{nameof(stepSize)} should be greater than zero");
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+
+namespace Algorithms.Numeric
+{
+    /// <summary>
+    /// In mathematics and computational science, the Euler method (also called forward Euler method)
+    /// is a first-order numerical procedure for solving ordinary differential equations (ODEs)
+    /// with a given initial value. It is the most basic explicit method for numerical integration
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:9]:
```suggestion
    /// with a given initial value (aka. Cauchy problem). It is the most basic explicit method for numerical integration
```

**Comment by algobytewise** [comment]
Thanks for the prompt and detailed feedback. I've implemented the changes suggested by you. I had to remove the test of the EulerStep-method, since it is private now. This change is responsible for the codecov/patch-error for #L78 - L79.

File: Algorithms.Tests/Numeric/EulerMethodTest.cs [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+{
+    public static class EulerMethodTest
+    {
+        [Test]
+        public static void TestLinearEquation()
+        {
+            Func<double, double, double> exampleEquation1 = (x, y) => x;
+            List<double[]> points =  Algorithms.Numeric.EulerMethod.EulerFull(0, 4, 0.1, 0, exampleEquation1);
+            double yEnd = points[points.Count - 1][1];
+            Assert.AreEqual(yEnd, 7.800000000000003);
```
Comment (siriak) [inline Algorithms.Tests/Numeric/EulerMethodTest.cs:16]: Why is it 7.8? And use yEnd.Should().BeApproximately(7.8).
y' = x
y = x^2 / 2 + C
Taking into account that y(0) = 0, C = 0
y = x^2 / 2
Then y(4) = 4^2 / 2 = 8, not 7.8

File: Algorithms.Tests/Numeric/EulerMethodTest.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        
+        [Test]
+        public static void TestExampleWikipedia()
+        {
+            // example from https://en.wikipedia.org/wiki/Euler_method
+            Func<double, double, double> exampleEquation2 = (x, y) => y;
+            List<double[]> points =  Algorithms.Numeric.EulerMethod.EulerFull(0, 4, 0.1, 1, exampleEquation2);
+            double yEnd = points[points.Count - 1][1];
+            Assert.AreEqual(yEnd, 45.25925556817596);
```
Comment (siriak) [inline Algorithms.Tests/Numeric/EulerMethodTest.cs:26]: From Wikipedia, error for this example is 9.34, which is huge as a relative error. Please use small steps to get correct answers.

File: Algorithms.Tests/Numeric/EulerMethodTest.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        
+        [Test]
+        public static void TestExampleGeeksForGeeks()
+        {
+            // example from https://www.geeksforgeeks.org/euler-method-solving-differential-equation/
+            Func<double, double, double> exampleEquation3 = (x, y) => x + y + x * y;
+            List<double[]> points =  Algorithms.Numeric.EulerMethod.EulerFull(0, 0.1, 0.025, 1, exampleEquation3);
+            double yEnd = points[points.Count - 1][1];
+            Assert.AreEqual(yEnd, 1.1116729841674804);
```
Comment (siriak) [inline Algorithms.Tests/Numeric/EulerMethodTest.cs:36]: Is this an analytical solution? We should only compare against it.

File: Algorithms.Tests/Numeric/EulerMethodTest.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        
+        [Test]
+        public static void TestExceptions()
+        {
+            Func<double, double, double> exampleEquation = (x, y) => x;
+            Assert.Throws<ArgumentOutOfRangeException>(() => Algorithms.Numeric.EulerMethod.EulerFull(0, 4, 0, 0, exampleEquation));
+            Assert.Throws<ArgumentOutOfRangeException>(() => Algorithms.Numeric.EulerMethod.EulerFull(0, -4, 0.1, 0, exampleEquation));
```
Comment (siriak) [inline Algorithms.Tests/Numeric/EulerMethodTest.cs:44]: Please have a separate test for each test case. Name them like StartAndEndCoincide_ThrowsArgumentOutOfRangeException, StartIsLargerThanEnd_ThrowsArgumentOutOfRangeException

File: Algorithms/Numeric/EulerMethod.cs [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yStart">Initial conditions y-value.</param>
+        /// <param name="differentialEquation">The right hand side of the differential equation.</param>
+        /// <returns>The solution of the Cauchy problem.</returns>
+        public static List<double[]> EulerFull(
+                double xStart,
+                double xEnd,
+                double stepSize,
+                double yStart,
+                Func<double, double, double> differentialEquation)
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:34]:
```suggestion
                Func<double, double, double> yDerivative)
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+        private static double EulerStep(
+                double xCurrent,
+                double stepSize,
+                double yCurrent,
+                Func<double, double, double> differentialEquation)
+        {
+            if (stepSize <= 0)
+            {
+              throw new ArgumentOutOfRangeException(nameof(stepSize), $"{nameof(stepSize)} should be greater than zero");
+            }
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:80]: This is a private method, no need for guard clauses

File: Algorithms/Numeric/EulerMethod.cs [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+        /// <param name="xCurrent">Current x-value.</param>
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yCurrent">Current y-value.</param>
+        /// <param name="differentialEquation">The differential equation to be solved.</param>
+        /// <returns>The next y-value.</returns>
+        private static double EulerStep(
+                double xCurrent,
+                double stepSize,
+                double yCurrent,
+                Func<double, double, double> differentialEquation)
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:75]:
```suggestion
                Func<double, double, double> yDerivative)
```

File: Algorithms/Numeric/EulerMethod.cs [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+            return points;
+        }
+
+        /// <summary>
+        /// Calculates the next y-value based on the current value of x, y and the stepSize.
+        /// </summary>
+        /// <param name="xCurrent">Current x-value.</param>
+        /// <param name="stepSize">Step-size on the x-axis.</param>
+        /// <param name="yCurrent">Current y-value.</param>
+        /// <param name="differentialEquation">The differential equation to be solved.</param>
```
Comment (siriak) [inline Algorithms/Numeric/EulerMethod.cs:69]:
```suggestion
        /// <param name="differentialEquation">The right hand side of the differential equation.</param>
```

File: Algorithms.Tests/Numeric/EulerMethodTest.cs (hunk 11 above)
Comment (algobytewise) [reply Algorithms.Tests/Numeric/EulerMethodTest.cs]: It should be 8 but that's the error due to the algorithm itself. I have decreased the step-size to get it closer.

File: Algorithms.Tests/Numeric/EulerMethodTest.cs (hunk 13 above)
Comment (algobytewise) [reply Algorithms.Tests/Numeric/EulerMethodTest.cs]: The idea of this test was not to test the precision of the algorithm but to make sure that the implementation is correct. I changed it to focus only on the first 2 steps and added comments to make it more clear what is happening.

Comment (siriak) [reply Algorithms.Tests/Numeric/EulerMethodTest.cs]: Ok, I see. This is a good idea too!

**Review by siriak — APPROVED** [review]
Looks good, thank you!

**Review by siriak — APPROVED** [review]
Could you add a link to README.md?

**Comment by algobytewise** [comment]
I've updated the readme. Thanks for all the helpful comments.

**Review by siriak — APPROVED** [review]

**Pull request merged**